File: backend/routers/interventions.py
"""
routers/interventions.py

Routes:
  POST   /interventions/generate           — bulk recommendation generation (RED/AMBER)
  GET    /interventions                    — list interventions (scoped by role)
  GET    /interventions/{id}              — get a single intervention
  PATCH  /interventions/{id}              — update status / notes / assignment
  DELETE /interventions/{id}              — remove (admin only)
"""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from loguru import logger

from modules.database import (  # patched as "routers.interventions.*" in tests
    db_create_alert,
    db_create_intervention,
    db_delete_intervention,
    db_get_employee_manager,
    db_get_employees_for_manager,
    db_get_intervention,
    db_get_interventions,
    db_get_latest_classifications,
    db_update_intervention,
    db_write_audit_log,
)
from modules.recommendations import generate_recommendation
from routers._helpers import build_employee_payload
from routers.deps import (
    GenerateRecommendationsRequest,
    LLMState,
    UpdateInterventionRequest,
    require_admin,
    require_any,
    require_manager_or_above,
)

router = APIRouter(tags=["interventions"])


@router.post("/interventions/generate")
async def generate_bulk_recommendations(
    body: GenerateRecommendationsRequest,
    user: dict = Depends(require_manager_or_above),
):
    """
    Generate personalised recommendations in bulk.

    Targets every RED/AMBER employee by default; pass ``employee_ids`` to
    narrow to specific people.
    """
    target_ids = body.employee_ids
    if not target_ids:
        zones          = body.zones or ["RED", "AMBER"]
        classifications = db_get_latest_classifications()
        target_ids     = [
            c["employee_id"] for c in classifications if c["risk_zone"] in zones
        ]

    if not target_ids:
        return {
            "status":        "ok",
            "generated":     0,
            "failed":        [],
            "alerts_created": 0,
            "results":       [],
            "message":       "No employees matched the target criteria.",
        }

    results        = []
    alerts_created = 0
    failed         = []

    for employee_id in target_ids:
        try:
            features, classification, recent_comments = build_employee_payload(employee_id)
        except ValueError as e:
            failed.append({"employee_id": employee_id, "error": str(e)})
            continue

        try:
            rec = generate_recommendation(
                employee_id=employee_id,
                features=features,
                classification=classification,
                recent_comments=recent_comments,
                llm=LLMState.instance,
            )
        except Exception as e:
            logger.error(f"[Recommendations] Bulk generation failed for {employee_id}: {e}")
            failed.append({"employee_id": employee_id, "error": str(e)})
            continue

        assigned_to = (
            db_get_employee_manager(employee_id) if body.assign_to_manager else None
        )

        intervention_id = db_create_intervention(
            employee_id=employee_id,
            created_by=user["username"],
            reasoning=rec["reasoning"],
            actions=rec["actions"],
            priority=rec["priority"],
            timeline=rec["timeline"],
            assigned_to=assigned_to,
        )

        if rec["priority"] in ("high", "critical"):
            db_create_alert(
                employee_id,
                "recommendation_generated",
                "critical" if rec["priority"] == "critical" else "warning",
                f"New {rec['priority']} priority recommendation generated for "
                f"{employee_id}: {rec['actions'][0]['title']}",
            )
            alerts_created += 1

        results.append({
            "intervention_id": intervention_id,
            "employee_id":     employee_id,
            "assigned_to":     assigned_to,
            **rec,
        })

    db_write_audit_log(
        user["username"],
        "generate_recommendations_bulk",
        "intervention",
        details=f"{len(results)} generated",
    )

    return {
        "status":        "ok",
        "generated":     len(results),
        "failed":        failed,
        "alerts_created": alerts_created,
        "results":       results,
    }


@router.get("/interventions")
async def list_interventions(
    employee_id: Optional[str] = Query(None),
    status: Optional[str] = Query(None),
    mine: bool = Query(False, description="Only interventions assigned to the current user"),
    limit: int = Query(100),
    user: dict = Depends(require_any),
):
    """
    List interventions.  Managers are automatically scoped to their direct
    reports; admins/HRBP see everything by default.
    """
    manager_scope = None
    assigned_to   = user["username"] if mine else None

    if not employee_id and user["role"] == "manager":
        manager_scope = db_get_employees_for_manager(user["username"])

    interventions = db_get_interventions(
        employee_id=employee_id,
        assigned_to=assigned_to,
        manager_employee_ids=manager_scope,
        status=status,
        limit=limit,
    )
    return {"interventions": interventions}


@router.get("/interventions/{intervention_id}")
async def get_intervention_by_id(intervention_id: int, _: dict = Depends(require_any)):
    intervention = db_get_intervention(intervention_id)
    if not intervention:
        raise HTTPException(status_code=404, detail="Intervention not found")
    return intervention


@router.patch("/interventions/{intervention_id}")
async def update_intervention(
    intervention_id: int,
    body: UpdateInterventionRequest,
    user: dict = Depends(require_manager_or_above),
):
    """Update status, notes, assignment, priority, or due date."""
    if not db_get_intervention(intervention_id):
        raise HTTPException(status_code=404, detail="Intervention not found")

    updated = db_update_intervention(
        intervention_id,
        status=body.status,
        notes=body.notes,
        assigned_to=body.assigned_to,
        priority=body.priority,
        due_date=body.due_date,
    )
    if not updated:
        raise HTTPException(status_code=400, detail="No fields provided to update")

    db_write_audit_log(user["username"], "update_intervention", "intervention", str(intervention_id))
    return {"status": "updated", "intervention_id": intervention_id}


@router.delete("/interventions/{intervention_id}")
async def delete_intervention(intervention_id: int, user: dict = Depends(require_admin)):
    if not db_delete_intervention(intervention_id):
        raise HTTPException(status_code=404, detail="Intervention not found")
    db_write_audit_log(user["username"], "delete_intervention", "intervention", str(intervention_id))
    return {"status": "deleted"}

File: frontend/lib/api/interventions.ts
/**
 * lib/api/interventions.ts
 * Frontend paths:  /interventions  |  /actions
 * Backend routes:  POST   /interventions/generate
 *                  GET    /interventions
 *                  GET    /interventions/{id}
 *                  PATCH  /interventions/{id}
 *                  DELETE /interventions/{id}
 */

import { BASE, authHeaders, handleResponse } from "./_core"

// ── Types ────────────────────────────────────────────────────────────────────

export interface RecommendationAction {
  title:       string
  owner:       string
  description: string
}

export interface Recommendation {
  reasoning: string
  priority:  "low" | "medium" | "high" | "critical"
  timeline:  string
  actions:   RecommendationAction[]
  source:    "llm" | "rule_based"
}

export interface Intervention extends Recommendation {
  id:           number
  employee_id:  string
  created_by:   string
  assigned_to:  string | null
  status:       string
  notes?:       string | null
  due_date?:    string | null
  created_at:   string
  updated_at?:  string
  completed_at?: string | null
}

export interface BulkRecommendationBody {
  employee_ids?:      string[]
  zones?:             string[]
  assign_to_manager?: boolean
}

export interface BulkRecommendationResult {
  status:        string
  generated:     number
  failed:        { employee_id: string; error: string }[]
  alerts_created: number
  results: ({
    intervention_id: number
    employee_id:     string
    assigned_to:     string | null
  } & Recommendation)[]
}

// ── Functions ────────────────────────────────────────────────────────────────

export async function generateBulkRecommendations(
  body: BulkRecommendationBody = {},
): Promise<BulkRecommendationResult> {
  const res = await fetch(`${BASE}/interventions/generate`, {
    method:  "POST",
    headers: { ...authHeaders(), "Content-Type": "application/json" },
    body:    JSON.stringify(body),
  })
  return handleResponse<BulkRecommendationResult>(res)
}

export async function listInterventions(params?: {
  employee_id?: string
  status?:      string
  mine?:        boolean
  limit?:       number
}): Promise<{ interventions: Intervention[] }> {
  const q = new URLSearchParams()
  if (params?.employee_id) q.set("employee_id", params.employee_id)
  if (params?.status)      q.set("status",      params.status)
  if (params?.mine)        q.set("mine",         "true")
  if (params?.limit)       q.set("limit",         String(params.limit))
  const res = await fetch(`${BASE}/interventions?${q}`, { headers: authHeaders() })
  return handleResponse<{ interventions: Intervention[] }>(res)
}

export async function getIntervention(id: number): Promise<Intervention> {
  const res = await fetch(`${BASE}/interventions/${id}`, { headers: authHeaders() })
  return handleResponse<Intervention>(res)
}

export async function updateIntervention(
  id:   number,
  body: {
    status?:      string
    notes?:       string
    assigned_to?: string
    priority?:    string
    due_date?:    string
  },
): Promise<{ status: string; intervention_id: number }> {
  const res = await fetch(`${BASE}/interventions/${id}`, {
    method:  "PATCH",
    headers: { ...authHeaders(), "Content-Type": "application/json" },
    body:    JSON.stringify(body),
  })
  return handleResponse<{ status: string; intervention_id: number }>(res)
}

export async function deleteIntervention(id: number): Promise<{ status: string }> {
  const res = await fetch(`${BASE}/interventions/${id}`, {
    method:  "DELETE",
    headers: authHeaders(),
  })
  return handleResponse<{ status: string }>(res)
}
